### PR TITLE
txn: check should_not_exist flag for Flush

### PR DESCRIPTION
**Pull request by ekexium**
### What is changed and how it works?

Issue Number: ref #16291

What's Changed:

```commit-message
Check should_not_exist flag for Flush, even when the lock is written by current transaction itself.

This is for the case:
TiDB flush key K.
TiDB do membuffer.CheckMemExist(K) and gets ErrNotFound because the key is flushed to TiKV. 
In vanilla TiDB it is able to know that the key has been written by membuffer.Get(K), 
but for pipelined DML it is too expensive to do a membuffer.Get because it involves a RPC.

To compensate, in this case we let TiDB attach a PresumeNotExist flag to the mutation.
In the next Flush, the should_not_exist flag is attached, 
and during the Flush we detect that the value is written before and return an AlreadyExist error.
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: src/storage/txn/actions/prewrite.rs [hunk 1]
```diff
@@ -446,6 +446,16 @@ impl<'a> PrewriteMutation<'a> {
             .into());
         }
 
+        // A key can be flushed multiple times for a Pipelined-DML transaction.
+        // A latter flush with `should_not_exist` should return error if a previous
+        // flush of the key writes a value
+        if self.should_not_exist && matches!(lock.lock_type, LockType::Put) {
```
Comment (cfzjywxk) [inline src/storage/txn/actions/prewrite.rs:452]: Is it possible to check if the lock is a `pipelined lock` type? It seems the check is unnecessary for normal transctions.

Comment (ekexium) [reply src/storage/txn/actions/prewrite.rs]: Makes sense

**Review by cfzjywxk — APPROVED** [review]

**Review by MyonKeminta — APPROVED** [review]

File: src/storage/txn/commands/flush.rs [hunk 2]
```diff
@@ -391,4 +412,17 @@ mod tests {
         must_commit(&mut engine, k, 1, 2);
         must_get(&mut engine, k, 3, v);
     }
+
+    #[test]
+    fn test_flushed_existence_check() {
+        let mut engine = TestEngineBuilder::new().build().unwrap();
+        let k = b"key";
+        let v = b"value";
+        must_flush_put(&mut engine, k, *v, k, 1, 1);
```
Comment (you06) [inline src/storage/txn/commands/flush.rs:421]: This case checks the existence of flushed lock.
What about adding a cacse which checks the existence of a committed txn's write record?

Comment (ekexium) [reply src/storage/txn/commands/flush.rs:466]: Done

**Pull request merged**